File: usr/scripts/check_artifacts.py
#!/usr/bin/env python3
"""Fail when catalog version.artifact paths are missing or untracked."""

from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path


def git_environment() -> dict[str, str]:
    return {
        key: value
        for key, value in os.environ.items()
        if key not in {"GIT_DIR", "GIT_WORK_TREE", "GIT_COMMON_DIR"}
    }


def catalog_artifact_paths(distribution_root: Path) -> list[Path]:
    packages = distribution_root / "v1" / "packages"
    paths: list[Path] = []
    for catalog in sorted(packages.rglob("*.json")):
        payload = json.loads(catalog.read_text())
        for version in payload.get("versions", []):
            relative = version.get("artifact")
            if relative:
                paths.append(distribution_root / relative)
    return paths


def tracked_paths(repo_root: Path) -> set[str] | None:
    probe = subprocess.run(
        ["git", "-C", str(repo_root), "rev-parse", "--is-inside-work-tree"],
        capture_output=True,
        env=git_environment(),
        check=False,
    )
    if probe.returncode != 0 or probe.stdout.decode().strip() != "true":
        return None
    listed = subprocess.run(
        ["git", "-C", str(repo_root), "ls-files", "-z"],
        capture_output=True,
        env=git_environment(),
        check=True,
    )
    return set(listed.stdout.decode().split("\0")) - {""}


def main() -> int:
    repo_root = Path(__file__).resolve().parents[2]
    distribution_root = repo_root / "prayers"
    if len(sys.argv) > 1:
        given = Path(sys.argv[1])
        distribution_root = given if given.is_absolute() else (Path.cwd() / given).resolve()
        repo_root = distribution_root.parent

    missing: list[str] = []
    untracked: list[str] = []
    tracked = tracked_paths(repo_root)

    for artifact in catalog_artifact_paths(distribution_root):
        try:
            display = str(artifact.relative_to(repo_root))
        except ValueError:
            display = str(artifact)
        if not artifact.is_file():
            missing.append(display)
            continue
        if tracked is not None and display not in tracked:
            untracked.append(display)

    if missing or untracked:
        for path in missing:
            print(f"missing: {path}", file=sys.stderr)
        for path in untracked:
            print(f"untracked: {path}", file=sys.stderr)
        print(
            "git add the listed .praypkg files, then rerun make check-artifacts; do not rerun make publish",
            file=sys.stderr,
        )
        return 1

    print("check-artifacts: ok")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: usr/scripts/check_artifacts.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "json"
require "open3"
require "pathname"
require "set"

# Host git wrappers set GIT_DIR; strip it so checks see the target repository.
GIT_ENVIRONMENT_EXCLUSIONS = %w[GIT_DIR GIT_WORK_TREE GIT_COMMON_DIR].freeze

def git_environment
  ENV.to_h.reject { |key, _| GIT_ENVIRONMENT_EXCLUSIONS.include?(key) }
end

def catalog_artifact_paths(distribution_root)
  packages = File.join(distribution_root, "v1", "packages")
  Dir.glob(File.join(packages, "**", "*.json")).sort.flat_map do |catalog|
    payload = JSON.parse(File.read(catalog, encoding: "UTF-8"))
    versions = payload["versions"] || []
    versions.filter_map do |version|
      relative = version["artifact"]
      File.join(distribution_root, relative) unless relative.nil? || relative.empty?
    end
  end
end

def tracked_paths(repo_root)
  stdout, _stderr, status = Open3.capture3(
    git_environment,
    "git", "-C", repo_root.to_s, "rev-parse", "--is-inside-work-tree"
  )
  return nil unless status.success? && stdout.strip == "true"

  listed, stderr, status = Open3.capture3(
    git_environment,
    "git", "-C", repo_root.to_s, "ls-files", "-z"
  )
  raise "git ls-files failed: #{stderr}" unless status.success?

  listed.split("\0").reject(&:empty?).to_set
end

def display_path(artifact, repo_root)
  Pathname.new(artifact).relative_path_from(Pathname.new(repo_root)).to_s
rescue ArgumentError
  artifact.to_s
end

def find_artifact_problems(distribution_root, repo_root)
  missing = []
  untracked = []
  tracked = tracked_paths(repo_root)

  catalog_artifact_paths(distribution_root).each do |artifact|
    display = display_path(artifact, repo_root)
    unless File.file?(artifact)
      missing << display
      next
    end
    untracked << display if tracked && !tracked.include?(display)
  end

  [missing, untracked]
end

def report_artifact_problems(missing, untracked, stdout: $stdout, stderr: $stderr)
  if missing.empty? && untracked.empty?
    stdout.puts "check-artifacts: ok"
    return 0
  end

  missing.each { |path| stderr.puts "missing: #{path}" }
  untracked.each { |path| stderr.puts "untracked: #{path}" }
  stderr.puts "git add the listed .praypkg files, then rerun make check-artifacts; do not rerun make publish"
  1
end

def resolve_roots(arguments, script_directory: __dir__, working_directory: Dir.pwd)
  if arguments.empty?
    repo_root = File.expand_path("../..", script_directory)
    [File.join(repo_root, "prayers"), repo_root]
  else
    distribution_root = File.expand_path(arguments.first, working_directory)
    [distribution_root, File.dirname(distribution_root)]
  end
end

if $PROGRAM_NAME == __FILE__
  distribution_root, repo_root = resolve_roots(ARGV)
  missing, untracked = find_artifact_problems(distribution_root, repo_root)
  exit report_artifact_problems(missing, untracked)
end
